### PR TITLE
Add MSIDURLSessionDelegate blocks

### PR DESCRIPTION
**Pull request by antrix1989**

### CODE REVIEW

File: IdentityCore/src/network/session_delegate/MSIDURLSessionDelegate.m [hunk 1]
```diff
@@ -25,4 +25,74 @@
...
+    MSID_LOG_INFO(nil, @"Redirecting to %@", _PII_NULLIFY(request.URL.absoluteString));
+    MSID_LOG_INFO_PII(nil, @"Redirecting to %@", request.URL.absoluteString);
+    
+    NSURLRequest *redirectRequest = request;
+    
+    if (self.taskWillPerformHTTPRedirectionBlock)
+    {
+        redirectRequest = self.taskWillPerformHTTPRedirectionBlock(session, task, response, request);
+    }
+
```
Comment (unpluggedk) [inline IdentityCore/src/network/session_delegate/MSIDURLSessionDelegate.m]: should we check scheme and only allow for https?

File: IdentityCore/src/network/session_delegate/MSIDURLSessionDelegate.m [hunk 2]
```diff
@@ -25,4 +25,74 @@
...
+    NSString *authMethod = [challenge.protectionSpace.authenticationMethod lowercaseString];
+    
+    MSID_LOG_VERBOSE(nil, @"%@ - %@. Previous challenge failure count: %ld", @"session:didReceiveChallenge:completionHandler", authMethod, (long)challenge.previousFailureCount);
+    
+    NSURLSessionAuthChallengeDisposition disposition = NSURLSessionAuthChallengePerformDefaultHandling;
+    __block NSURLCredential *credential = nil;
+    
+    if (self.sessionDidReceiveAuthenticationChallengeBlock)
+    {
+        disposition = self.sessionDidReceiveAuthenticationChallengeBlock(session, challenge, &credential);
```
Comment (oldalton) [inline IdentityCore/src/network/session_delegate/MSIDURLSessionDelegate.m]: nit: This is a bit different from how interactive challenge handlers are implemented, where challenge handler calls the completionHandler itself. Can we make sure they both behave the same? I have no personal preference over one or another.

File: IdentityCore/src/network/session_delegate/MSIDURLSessionDelegate.m [hunk 3]
```diff
@@ -25,4 +25,74 @@
 
 @implementation MSIDURLSessionDelegate
 
+#pragma mark - NSURLSessionDelegate
+
+- (void)URLSession:(NSURLSession *)session didReceiveChallenge:(NSURLAuthenticationChallenge *)challenge
```
Comment (oldalton) [inline IdentityCore/src/network/session_delegate/MSIDURLSessionDelegate.m]: just curious, why do we need both this one and the URLSession:task:... methods?
Apple documentation says: `"If you do not implement this method, the session calls its delegate’s URLSession:task:didReceiveChallenge:completionHandler: method instead"`

File: IdentityCore/src/network/session_delegate/MSIDURLSessionDelegate.m (hunk 1 above)
Comment (oldalton) [reply IdentityCore/src/network/session_delegate/MSIDURLSessionDelegate.m]: That should be the responsibility of whoever implements the blocks, I assume.

**Pull request merged**